Client/Frontend/Browser/Search/BraveSearchManager.swift

```diff
@@ -0,0 +1,292 @@
+        profile.logins.getLoginsForProtectionSpace(challenge.protectionSpace) >>== { cursor in
+            guard cursor.count >= 1 else {
+                completion(nil)
+                return
+            }
+
+            let logins = cursor.asArray()
+            var credentials: URLCredential?
+
+            if logins.count > 1 {
+                credentials = (logins.find { login in
+                    (login.protectionSpace.`protocol` == challenge.protectionSpace.`protocol`) && !login.hasMalformedHostname
+                })?.credentials
+            } else if logins.count == 1, logins.first?.protectionSpace.`protocol` != challenge.protectionSpace.`protocol` {
+                credentials = logins.first?.credentials
+            } else {
+                credentials = logins.first?.credentials
+            }
+            
+            completion(credentials)
+        }
+    }
+    
+    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+        
+        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodHTTPBasic ||
+              challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodHTTPDigest ||
+              challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodNTLM else {
+            completionHandler(.performDefaultHandling, nil)
+            return
+        }
+        
+        let validURLs = AppConstants.buildChannel.isPublic ?
+            ["search.brave.com"] : ["search.brave.com", "search-dev.brave.com"]
+        if !validURLs.contains(challenge.protectionSpace.host) {
+            completionHandler(.performDefaultHandling, nil)
+            return
+        }
+        
+        // -- Handle Authentication --
+        
+        // Too many failed attempts
+        if challenge.previousFailureCount >= 3 {
+            completionHandler(.rejectProtectionSpace, nil)
+            return
+        }
+        
+        if let credentials = BraveSearchManager.cachedCredentials {
+            completionHandler(.useCredential, credentials)
+            return
+        }
+
+        if let proposedCredential = challenge.proposedCredential,
+           !(proposedCredential.user?.isEmpty ?? true),
+           challenge.previousFailureCount == 0 {
+            completionHandler(.useCredential, proposedCredential)
+            return
+        }
+
+        // There is only ever ONE profile and all tabs share it afaict
+        let profile = { () -> Profile? in
+            if Thread.current.isMainThread {
+                return (UIApplication.shared.delegate as? AppDelegate)?.browserViewController.profile
+            }
+            
+            return DispatchQueue.main.sync {
+                (UIApplication.shared.delegate as? AppDelegate)?.browserViewController.profile
+            }
```

Comment:
This will cause a deadlock because your delegate queue is `.main`. This code doesn't seem like it will ever execute though so perhaps its best to simply remove

Reply:
@Brandon-T why is it even main thread ensured like that?